src/main/java/com/sps/Servlets/NewRecipeServlet.java

```diff
@@ -0,0 +1,146 @@
+      recipeEntity = datastore.get(KeyFactory.stringToKey(keyString));
+    } catch (EntityNotFoundException e) {
+      e.printStackTrace();
+      return;
+    }
+    Recipe original = entityToRecipe(recipeEntity);
+    response.setContentType("application/json;");
+    response.getWriter().println(convertToJsonUsingGson(original));  
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Collection<String> searchStrings = new HashSet<>();
+    String name = request.getParameter("name");
+    searchStrings.add(name.toUpperCase());
+    String description = request.getParameter("description");
+    Collection<EmbeddedEntity> tags = getParameters(request, TAG, searchStrings);
+    Collection<EmbeddedEntity> ingredients = getParameters(request, INGREDIENT, searchStrings);
+    Collection<EmbeddedEntity> steps = getParameters(request, STEP, null);
+
+    Entity recipe = new Entity("Recipe");
+    recipe.setProperty("name", name);
+    recipe.setProperty("description", description);
+    recipe.setProperty("tags", tags);
+    recipe.setProperty("ingredients", ingredients);
+    recipe.setProperty("steps", steps);
+    recipe.setProperty("search-strings", new ArrayList<String>(searchStrings));
+    datastore.put(recipe);
+
+    response.sendRedirect("/edit-recipe.html");
+  }
+
+  /**
+   * Gets the parameters for fields that have different numbers of parameters from recipe to recipe.
+   * For example, one recipe may have 2 ingredients, while another may have 20.
+   * This method ensures that all tags, ingredients, and steps are recorded, no matter how many of each a recipe has.
+   */
+  private Collection<EmbeddedEntity> getParameters(HttpServletRequest request, String type, Collection<String> searchStrings) {
+    Collection<EmbeddedEntity> parameters = new LinkedList<>();
+    int parameterNum = 1;
+
+    String parameterName = type + parameterNum;
+    String parameter = request.getParameter(parameterName);
+    while (parameter != null) {
+      EmbeddedEntity parameterEntity = new EmbeddedEntity();
+      parameterEntity.setProperty(type, parameter);
+      addToSearchStrings(searchStrings, parameter);
+      parameterName = type + (++parameterNum);
+      parameter = request.getParameter(parameterName);
+      parameters.add(parameterEntity);
+    }
+    return parameters;
+  }
+
+  /**
+   * Adds a formatted search string to the set of search strings.
+   */
+  private void addToSearchStrings(Collection<String> searchStrings, String stringToAdd) {
+    if (searchStrings == null) {
+      return;
+    }
+    searchStrings.add(stringToAdd.toUpperCase());
+  }
+
+  private Recipe entityToRecipe(Entity recipeEntity) {
+    String name = (String) recipeEntity.getProperty("name");
+    String description = (String) recipeEntity.getProperty("description");
+    LinkedList<String> tags = (LinkedList<String>) (LinkedList<?>) getDataAsList(recipeEntity.getProperty("tags"), TAG);
```

Comment:
Thing is interesting to me, because when I tried to use Gson to convert linked lists into json, it didn't work, and the internet claimed that only Arraylists work immediately with Gson. Did you confirm that it actually converts the linked lists properly?

Reply:
I assume and hope that it works properly, because I never ran into any problems as I tested it.